doc/grpc_client/design_doc.md

```diff
@@ -0,0 +1,111 @@
+## gRPC client for active-actve DualToR scenario design
+
+## Revision
+
+| Rev | Date     | Author          | Change Description |
+|:---:|:--------:|:---------------:|--------------------|
+| 0.1 | 04/1/22 | Vaibhav Dahiya  | Initial version    |
+
+## Scope
+
+### Overview
+
+This document summarizes the approach taken to accommodate gRPC client
+for DualToR active active scenario. The gRPC client daemon's
+main purpose is to provide a way for linkmgr to exchange RPC's with SoC
+and do this within SONiC PMON docker
+
+
+Challenge: to provide an interface for gRPC daemon to exchange RPC's with the gRPC server running on the SoC
+           over a secure channel using a loopback IP. The RPC's exchanged with SoC
+           would help linkmgr state machine make decisions as to transition the DualToR
+           into active/standby depending on the state of the SoC 
+
+
+### gRPC client communicate to SoC over Loopback IP
+
+#### Background
+
+- We need a way to communicate to SoC using a Loopback IPv4 which is not adversitised to public. This Loopback IP is a well defined IP which is allowed in SoC firewall rules, hence
+the requirement to communicate over a pre-defined IPv4 address. 
+
+  - The Best approach would be BIND the socket using the gRPC to Loopback interface using gRPC API. The socket_mutator API is available in C++, which allows us to accomplish this. However the gRPC client is prposed Python daemon run inside PMON container. As such as long as library does not expose this API in Python, this is not an easy workaround. The github issue is filed for gRPC library.
+  - Adding a Kernel Route. We could add a Kernel Route to the soc IP. For example
+    ```
+       sudo ip route add <soc IP> via <vlan IP> src <Loopback IP>
+    ```
+  - Te issue with adding a Kernel Route is the route_cheker will fail for this, since vlan IP is the HOST's own vlan IP within SONiC as such no real neighbor is present
+  - SWSS orchagent will complain about not able to istall the entry in ASIC, the entry will be present in APP DB but not present inside ASIC. This would deem more worarounds necessary to be able to use this approach.
+  - using IPTABLES rule. For Example
+    ```
+        sudo iptables -t nat -A POSTROUTING --destination <soc IP> -j SNAT --to-source <LoopBack IP>
+    ```
+
+#### Proposed Solution
+
+- use the IPTABLES rule approach as with this approach, this workaround is cleaner in a sense there are no more workarounds after adding the rule. Caclmgrd will check the CONFIG DB DEVICE_METADATA and upon learning this is ToR with subtype DualToR, will add the IPTABLES rule.
+    ```
+        DEVICE_METADATA | localhost
+        type: ToRRouter
+	subtype: DualToR 
+    ```
+    ```
+        LINKMGR_CABLE|PORTNAME
+        soc_ipv4: <soc IP>
+    ```
+- The update_control_plane_nat_acls in caclmgrd will look for the above configuration and upon getting the config, it will add the POSTROUTING nat rules
```

Comment:
Please specify what NAT rules are installed currently and if it has any dependencies. How about restarting caclmgrd? Please capture in the doc

Reply:
fixed